### ISSUE TITLE
The same filter update by SetFilter method enables full cache load

### ISSUE THREAD

**Issue by vicshulha** [cgrates/cgrates#4357]
Hi All,

Upon recommendation from Ionuț at https://groups.google.com/g/cgrates I am creating the issue report here.

We face a new issue loading the full Redis cache during SetFilter method usage. It started to happen after CGRates version update from CGR...@v0.10.2~dev-20200831063159-3289a59a41c3 to the latest one (CGR...@v0.10.3).

The issue symptoms and negative effects:

Redis cache load generates huge network traffic (1.5G data sent by 60mbit/s for 2 min.) between CGrates and Redis nodes. In turn, this behavior leads to CGRates performance degradation which fails with “ATTRIBUTES_ERROR:context deadline exceeded” error and does not allow processing a part of new requests for some time.

Steps to reproduce (STR) the issue:

1) Create attribute profile (SetAttributeProfile.txt);
2) Set a filter to add 1k numbers (SetFilter_1.txt);
3) Set the same filter with 1 number to delete the rest of 999 ones (SetFilter_2.txt);

After the 3rd step, the issue symptoms become observable namely huge network traffic starts, and CGRates is loaded (see CPU statistics). If to add 1k numbers again (step 2) after the STR, no cache load process has started but only reappears after step 3.

Please find the files used during STR uploaded to GDrive (https://drive.google.com/drive/folders/1cvlzfYa0rKQQM3cxwvojiKM0Py0E4FX0?usp=drive_link):

- ngrep log (6379,2080,2012 ports);
- requests used in STR;
- cgrates.json, cgrates.cfg (kamailio);
- network, CPU statistics screen for 2 min.

I would appreciate an idea of what CGRates change led to this behavior and how it can be fixed.

**Comment by ionutboangiu** [comment]
Hi,

I tried to follow the steps you provided but couldn't reproduce the issue. 

1.  The Set* API requests you sent reload the cache by default unless configured otherwise. This is determined by the "default_caching" setting in the JSON configuration file (can be overwritten by the cache specific request parameter). Since this setting wasn't in your configuration, the default *reload behavior happened.
    
2.  When I followed your steps, the process was quick for me because I don't have much cached data.
    
3.  It's normal to see higher network traffic during the setup phase as the cache is being filled/reloaded and the indexes are being built (as long as it doesn't break the system). This is to make sure outdated data is not being used.
    
Could you please tell me:

-   How long did you experience the high network traffic?
-   Did you have any performance issues after the cache reload finished?

Additionally, could you provide the output of either the following cgr-console command:
```
cgr-console cache_stats
```
or the output of this equivalent JSON request:
```
{"method":"CacheSv1.GetCacheStats","params":[{"Tenant":"","CacheIDs":null}],"id":1}
```

Could you also try to reproduce the issue from scratch and give me the complete steps so I can attempt to reproduce it myself?

Thanks,
Ionuț

**Comment by vicshulha** [comment]
Hi Ionuț,

Thank you for your response.

Please find my replies to your statements:

> 1. The Set* API requests you sent reload the cache by default unless configured otherwise. This is determined by the 
> "default_caching" setting in the JSON configuration file (can be overwritten by the cache specific request parameter). 
> Since this setting wasn't in your configuration, the default *reload behavior happened.

Yes, we know that there are Remove, Clear, Load, Reload values and Reload should be as default one. However, it looks like all the cache is loaded instead of some delta only by Reload mode.

> 2. When I followed your steps, the process was quick for me because I don't have much cached data.

Probably, I provided the wrong order of requests. I will provide the corrected STR further. I retested with a clean Redis cache and reached some gaps also. Just for your information, here is memory usage statistics:

- Production Redis (RAM: 6.38 GiB)
4% of memory is used (256 MB).

- Staging Redis (RAM: 6.38 GiB)
1.75% of memory is used (112 MB).

> 3. It's normal to see higher network traffic during the setup phase as the cache is being filled/reloaded and the indexes > are being built (as long as it doesn't break the system). This is to make sure outdated data is not being used.

I see this point however according to network load it looks like all the data is reloaded instead of delta. 

> Could you please tell me:
> How long did you experience the high network traffic?
> Did you have any performance issues after the cache reload finished?

Within the STR which I will provide below, I have received the following timings after SetFilter method triggered the issue.

- With clean Redis cache - 4 sec.
`real	0m4.244s
user	0m0.000s
sys	0m0.006s`
- Staging system - 23 sec.
`real	0m23.838s
user	0m0.005s
sys	0m0.002s`
- Production system - about 2-3 min. periodically, however, there was the case with 3h duration at peak traffic hours. Can not provide the exact timing example as can not test in production. It looked like there was some loop when cache load could not be completed probably due to load with other requests.

So, providing you with the STR with exact requests and their order tested with both Staging Redis and clean one:

1. Set filter_#1 to add 1k numbers (SetFilter_1.txt);
2. Set filter_#2 to add 1k numbers (SetFilter_2.txt); 
4. Create attribute profile with reference to filter_#1 and filter_#2 (SetAttributeProfile.txt);
5. Set a filter_#1 with 1 number to delete the rest of 999 ones (SetFilter_3.txt);
6. 4 sec, 23 sec, and 3 min gap (network load) is observed with clean, staging and production environment apropriately.

Please find the files used during STR uploaded to GDrive (https://drive.google.com/drive/folders/1emivtD7jMDzYv9iJhWfasP86_7zemey8?usp=sharing):

- requests used in STR;
- cache statistics output after STR with clean and staging Redis.

Please let me know if you managed to reproduce the issue or if additional steps from my end are needed.

**Comment by ionutboangiu** [comment]
Thanks again for the detailed steps. You've made it very easy to reproduce the issue.

It’s now clear that most of the network traffic is from removing outdated filter indexes after updating the filter. In version v0.10.2, these old indexes would remain in the datadb, potentially causing other issues.

Please keep this issue open and we will analyze if we can optimize the process.

**Comment by vicshulha** [comment]
Hi Ionuț,

Thank you for the description of the difference between versions and will wait for possible recommendations to isolate the negative effect on our current production traffic.

**Comment by vicshulha** [comment]
Hi Ionut,

While the described above issue is being analyzed from your end, we tried to isolate the negative consequences of delayed CGRates replies by kamailio-agent failover to the 2nd node, however, found the process does not work as expected.

The process looks as follows:

At the time of the initially described issue calls failed with "Error":"ATTRIBUTES_ERROR:context deadline exceeded" error which kamailio-agent sends to kamailio when can not receive response from CGRates during a particular interval ("reply_timeout": "3s").

We assume the error also happens due to kamailiio-agent can not connect to an alternative rpc connection (the 2nd GCRates node) to failover the loaded one. 
1) At the time of the issue, kamailio-agent tries to cancel a particular session with hung CGrates: `{"method":"_goRPC_.Cancel","params":[{"Seq":6444060}],"id":6444146}`
2) However, CGRates responds with the non-existent method and probably does not allow the agent to start the failover process.
`{"id":6444146,"result":null,"error":"rpc: can't find service _goRPC_.Cancel"}` 

Here are original requests and responses flow collected on the agent and the 1st CGRates node:

```
T 2024/05/28 10:31:31.820053 127.0.0.1:46932 -> 127.0.0.1:2012 [AP] #145108
{"method":"AttributeSv1.ProcessEvent","params":[{"AttributeIDs":null,"Context":"*sessions","ProcessRuns":null,"Tenant":"wildix.com","ID":"cbc6518","Time":"2024-05-28T10:31:31Z","Event":{"Account":"XXXX_UK","Category":"inbound_call","Destination":"441221234567","EvapiConnID":0,"OriginHost":"127.0.0.1:8448","RequestType":"*rated","SetupTime":"1716892291","Source":"KamailioAgent","Subject":"441221234567","Tenant":"wildix.com","cgr_context":"routing"}}],"id":6444060}
axCostStrategy":

T 2024/05/28 10:31:34.820366 127.0.0.1:46932 -> 127.0.0.1:2012 [AP] #145433
{"method":"_goRPC_.Cancel","params":[{"Seq":6444060}],"id":6444146}
com","Subject":"

#
T 2024/05/28 10:31:34.820458 127.0.0.1:59594 -> 127.0.0.1:8448 [AP] #145434
241:{"Event":"XXXX_ROUTING_REPLY","TransactionIndex":"44334","TransactionLabel":"116557686","Attributes":"","ResourceAllocation":"","MaxUsage":0,"Suppliers":"","Thresholds":"","StatQueues":"","Error":"ATTRIBUTES_ERROR:context deadline exceeded"},international","

#
T 2024/05/28 10:31:34.820980 127.0.0.1:2012 -> 127.0.0.1:46932 [AP] #145435
{"id":6444146,"result":null,"error":"rpc: can't find service _goRPC_.Cancel"}
```

How we can fix the above behavior of CGRates to at least achieve a failover process while the initial issue is being investigated?

**Comment by ionutboangiu** [comment]
Hello,

This issue should have been fixed in the latest v0.10. Please let me know if you still encounter any problems. 

Thank you once again for the find,
Ionuț

**Comment by vicshulha** [comment]
Hi Ionuț,

Thank you for the provided fix. We will test it soon and provide you with the feedback.

**Comment by vicshulha** [comment]
Hi Ionuț,

> This issue should have been fixed in the latest v0.10. Please let me know if you still encounter any problems.

We have not found an updated .deb package in http://pkg.cgrates.org/deb/v0.10/ therefore decided to build CGRateS@v0.10.4~dev-20240610165520-cc7210bbeee7 from sources by referring to the following instruction on Debian 11 GNU host:

https://cgrates.readthedocs.io/en/v0.10/installation.html#installing-from-source

But on the stage of building cgrates-*.deb package by 'make deb' command we face the following issue:
```
~/go/src/github.com/cgrates/cgrates/packages# make deb
...
dpkg-genchanges: info: including full source code in upload
E: cannot canonicalize filename /var/cache/pbuilder/base-bookworm+go.cow, does not exist
Undoing changes
Updated 1 path from the index
~/go/src/github.com/cgrates/cgrates/packages# cat /etc/issue
Debian GNU/Linux 11 \n \l
~/go/src/github.com/cgrates/cgrates/packages# uname -a
Linux redis-server 5.10.0-27-cloud-amd64 #1 SMP Debian 5.10.205-2 (2023-12-31) x86_64 GNU/Linux
```

Could you please share any hints about the requirements for the building environment or upload the latest v0.10 package to the repository?

**Comment by ionutboangiu** [comment]
Hello,

You can find the latest v0.10 debian package [here](http://pkg.cgrates.org/deb/v0.10/bullseye/cgrates_0.10.4~dev%2B20240610185520%2Bcc7210bbe-1~deb11u1_amd64.deb).

Thanks,
Ionuț

**Comment by vicshulha** [comment]
Hi Ionuț,

We have installed the package with the fix and the initial issue is no longer observed. Delete operation by SetFilter method is executed in 2ms instead of 90+ sec. previously.

However, the error observed with failover to the 2nd node mentioned in https://github.com/cgrates/cgrates/issues/4357#issuecomment-2145552319 is still existing. Here is the flow description:

Preconditions:
- There are 2 CGRates nodes configured in rpc connections of the agent;
- Failover timeout to the 2nd node - 3 sec.;
- The 1st node is loaded intentionally by stress test requests.

1) Agent send auth request to the 1st node
```
...
T 2024/06/17 08:59:04.352764 127.0.0.1:31164 -> 127.0.0.1:2012 [AP] #84
{"method":"AttributeSv1.ProcessEvent","params":[{"AttributeIDs":null,"Context":"*sessions","ProcessRuns":null,"Tenant":"wildix.com","ID":"a5f5b85","Time":"2024-06-17T08:59:04Z","Event":{"Account":"221100111111","Destination":"380711111111","EvapiConnID":0,"OriginHost":"127.0.0.1:8448","Partner":"\u003cnull\u003e","Role":"\u003cnull\u003e","SetupTime":"1718614744","Source":"KamailioAgent","SourceNumber":"393211111111","Subject":"393211111111","Tenant":"wildix.com","cgr_computelcr":"true","cgr_context":"auth"}}],"id":1}
on:6.06.20240213
#
...
```
2) The 1st node does not respond during 3 sec. therefore, the agent cancels the previous transaction and sends the request to the failover node
```
...
 T 2024/06/17 08:59:07.353732 127.0.0.1:31164 -> 127.0.0.1:2012 [AP] #87
{"method":"_goRPC_.Cancel","params":[{"Seq":1}],"id":2}
................
#
T 2024/06/17 08:59:07.353748 127.0.0.1:2012 -> 127.0.0.1:31164 [A] #88
{"id":6,"result"
#
T 2024/06/17 08:59:07.353839 10.47.2.227:18678 -> 18.158.245.207:2012 [AP] #89
{"method":"AttributeSv1.ProcessEvent","params": ...
...
```
3) The 1st node responds to the agent that it does not know anything about "Cancel" method 
```
...
#
T 2024/06/17 08:59:07.353998 127.0.0.1:2012 -> 127.0.0.1:31164 [AP] #90
{"id":2,"result":null,"error":"rpc: can't find service _goRPC_.Cancel"}
ntext":"*session
#
...
```

Could you please let us know why this behavior of an unknown method/service happens? If this error that should be branched as a new bug report please let me know. Thanks.
